### DISCUSSION TITLE
Self-hosting

### DISCUSSION THREAD

**Discussion by Villhelm-E** [PeWu/topola-viewer#61] [Q&A]
Is it possible to run topola on my personal website and password protect it, so I can share with family?

**Comment by PeWu** [comment]
Yes, you can run topola on your own server.

You can build the app from source by running `npm run build` in the cloned repository. Then, you can put the contents of the created `build/` directory on your server.

You can also find a ready code package in the `gh-pages` branch on GitHub. A downloadable package is here: https://github.com/PeWu/topola-viewer/archive/refs/heads/gh-pages.zip

**Reply by Villhelm-E** [reply]
I tried running `npm run build` in the cloned repo and I got this error:

<img width="566" alt="image" src="https://user-images.githubusercontent.com/25539368/125861248-bd1a7447-edd9-47c5-85d9-d86bb798ec06.png">

So then I downloaded that package you linked and it seems to work, but I don't know how to get it to automatically load my family tree. Could you help me here?

**Reply by jonezrin** [reply]
I also downloaded the package, it works, but only by uploading a file. Loading from URL does not work. Any chance of updating/fixing that? Or steering me toward a solution? Thanks.

**Reply by PeWu** [reply]
@jonezrin,

Simple answer: Try adding `&handleCors=false` to the URL when self-hosting.

More details: The problem is that some servers restrict the Web pages that can load a specific file by URL. By default, Topola Viewer uses a [CORS anywhere](https://github.com/Rob--W/cors-anywhere) proxy hosted at https://topola-cors.herokuapp.com/ which is restricted to Topola Viewer hosted on Github Pages. This means that you cannot use this proxy when you run self-hosted Topola Viewer. Disabling the CORS proxy by adding the `&handleCors=false` URL parameter is the easiest way to not use the proxy. If it turns out the Gedcom URL does have CORS restrictions, you may have to either change the server configuration where the Gedcom file is hosted or run a CORS proxy yourself.

**Reply by jonezrin** [reply]
Simple answer: It worked!!

More details: Thank you for replying and for your excellent coding.

**Reply by jonezrin** [reply]
Hi. Only problem I'm having now is the page not always loading the latest version of the GED file if updates are made to it. I tried no-cache metas in the index page but no luck. Any suggestion? Thanks.

**Reply by PeWu** [reply]
Topola Viewer has a built-in cache (sessionStorage). Reopen everything in a new tab to reset the cache.
